src/segment-loader.js

```diff
@@ -1527,11 +1478,53 @@ export default class SegmentLoader extends videojs.EventTarget {
       part,
       byteLength: 0,
       transmuxer: this.transmuxer_,
-      audioAppendStart,
-      gopsToAlignWith
+      // type of getMediaInfoForTime that was used to get this segment
+      getMediaInfoForTime
     };
+
+    const overrideCheck =
+      typeof forceTimestampOffset !== 'undefined' ? forceTimestampOffset : this.isPendingTimestampOffset_;
+
+    segmentInfo.timestampOffset = this.timestampOffsetForSegment_({
+      segmentTimeline: segment.timeline,
+      currentTimeline: this.currentTimeline_,
+      startOfSegment,
+      buffered: this.buffered_(),
+      overrideCheck
+    });
+
+    // if timestampoffset was set then we no longer have a timestampoffset
+    if (typeof segmentInfo.timestampOffset === 'number') {
+      this.isPendingTimestampOffset_ = false;
```

Comment:
It may be worth moving this out of the function too to make the function have no side effects. I think `generateSegmentInfo` will only be used on `chooseNextRequest` (as part of `fillBuffer`), which has a check for `timestampOffset` right after, and in `loadSegment_` as part of the queue callback, where it was forced to update.